backend/de.metas.contracts/src/main/java/de/metas/contracts/repository/impl/SubscriptionDiscountRepository.java

```diff
@@ -0,0 +1,121 @@
+		final IQueryBuilder<I_C_SubscrDiscount> subscrDiscountIdQuery = queryBL.createQueryBuilder(I_C_Flatrate_Conditions.class)
+				.addEqualsFilter(I_C_Flatrate_Conditions.COLUMNNAME_C_Flatrate_Conditions_ID, query.getFlatrateConditionId())
+				.andCollect(I_C_Flatrate_Conditions.COLUMN_C_SubscrDiscount_ID, I_C_SubscrDiscount.class);
+
+		final IQuery<I_M_Product> productQuery = queryBL.createQueryBuilder(I_M_Product.class)
+				.addOnlyActiveRecordsFilter()
+				.addEqualsFilter(I_M_Product.COLUMNNAME_M_Product_ID, query.getProductId())
+				.create();
+
+		final I_C_SubscrDiscount_Line lowestSeqNoLine = queryBL.createQueryBuilder(I_C_SubscrDiscount_Line.class)
+				.addInSubQueryFilter(I_C_SubscrDiscount_Line.COLUMNNAME_C_SubscrDiscount_ID, I_C_Flatrate_Conditions.COLUMNNAME_C_SubscrDiscount_ID, subscrDiscountIdQuery.create())
+				.addOnlyActiveRecordsFilter()
+				.filter(queryBL.createCompositeQueryFilter(I_C_SubscrDiscount_Line.class)
+						.addFilter(queryBL.createCompositeQueryFilter(I_C_SubscrDiscount_Line.class)
+								.addEqualsFilter(I_C_SubscrDiscount_Line.COLUMNNAME_M_Product_ID, null)
+								.addInSubQueryFilter(I_C_SubscrDiscount_Line.COLUMNNAME_M_Product_ID, I_M_Product.COLUMNNAME_M_Product_ID, productQuery)
+								.setJoinOr()
+						)
+						.addFilter(queryBL.createCompositeQueryFilter(I_C_SubscrDiscount_Line.class)
+								.addEqualsFilter(I_C_SubscrDiscount_Line.COLUMNNAME_M_Product_Category_ID, null)
+								.addInSubQueryFilter(I_C_SubscrDiscount_Line.COLUMNNAME_M_Product_Category_ID, I_M_Product.COLUMNNAME_M_Product_Category_ID, productQuery)
+								.setJoinOr()
+						)
+				)
+				.filter(queryBL.createCompositeQueryFilter(I_C_SubscrDiscount_Line.class)
+						.setJoinOr()
+						// startMonth null
+						.addFilter(queryBL.createCompositeQueryFilter(I_C_SubscrDiscount_Line.class)
+								.addEqualsFilter(I_C_SubscrDiscount_Line.COLUMNNAME_StartMonth, null)
+						)
+						// OR startMonth < given month
+						.addCompareFilter(I_C_SubscrDiscount_Line.COLUMNNAME_StartMonth, CompareQueryFilter.Operator.LESS_OR_EQUAL, month)
+						// OR
+						.addFilter(queryBL.createCompositeQueryFilter(I_C_SubscrDiscount_Line.class)
+								// startMonth = given month
+								.addEqualsFilter(I_C_SubscrDiscount_Line.COLUMNNAME_StartMonth, month)
+								// AND
+								.addFilter(queryBL.createCompositeQueryFilter(I_C_SubscrDiscount_Line.class)
+										.setJoinOr()
+										// startDay null
+										.addEqualsFilter(I_C_SubscrDiscount_Line.COLUMNNAME_StartDay, null)
+										// OR startDay <= given day
+										.addCompareFilter(I_C_SubscrDiscount_Line.COLUMNNAME_StartDay, CompareQueryFilter.Operator.LESS_OR_EQUAL, dayOfMonth)))
+
+				)
+				.orderBy(I_C_SubscrDiscount_Line.COLUMNNAME_SeqNo)
+				.create()
+				.first();
+
+		return lowestSeqNoLine == null ? null : from(lowestSeqNoLine);
+	}
+
+	private SubscriptionDiscount from(final I_C_SubscrDiscount_Line lineFromDB)
```

Comment:
minor: instead of `lineFromDB` u can call it `record`.

Reply:
renamed